lib/runners/processor/clang_tidy.rb

```diff
@@ -0,0 +1,107 @@
+
+    # @see https://github.com/github/linguist/blob/775b07d40c04ef6e0051a541886a8f1e30a892f4/lib/linguist/languages.yml#L532-L535
+    # @see https://github.com/github/linguist/blob/775b07d40c04ef6e0051a541886a8f1e30a892f4/lib/linguist/languages.yml#L568-L584
+    VALID_EXTENSIONS = [".c", ".cpp", ".c++", ".cc", ".cp", ".cxx"].freeze
+    GLOB_HEADERS = "**/*.{h,h++,hh,hpp,hxx,inc,inl,ipp,tcc,tpp}".freeze
+
+    def setup
+      deploy_packages
+      yield
+    end
+
+    def analyzer_bin
+      "clang-tidy"
+    end
+
+    def analyze(changes)
+      issues = []
+
+      changes
+        .changed_paths
+        .select { |path| VALID_EXTENSIONS.include?(path.extname.downcase) }
+        .map{ |path| relative_path(working_dir / path, from: current_dir) }
+        .reject { |path| path.to_s.start_with?("../") } # reject files outside the current_dir
+        .each do |path|
+          stdout, = capture3!(analyzer_bin, path.to_s, "--", *option_include_path,
+            is_success: ->(status) { [0, 1].include?(status.exitstatus) })
+          ret = construct_result(stdout)
+          issues.push(*ret)
+        end
+
+      Results::Success.new(guid: guid, analyzer: analyzer).tap do |result|
+        result.add_issue(*issues)
+      end
+    end
+
+    private
+
+    def deploy_packages
+      # select "lib*" and report others as warning for security concerns
+      packages = Array(config_linter[:apt])
+        .select do |pkg|
+          if pkg.start_with?("lib")
+            true
+          else
+            add_warning "Installing the package `#{pkg}` is blocked."
+            false
+          end
+        end
+
+      unless packages.empty?
+        capture3!("sudo", "apt-get", "install", "-qq", "-y", "--no-install-recommends", *packages)
+      end
+    end
+
+    def construct_result(stdout)
+      issues = []
+
+      stdout.each_line do |line|
+        # issue format
+        # <path>:<line>:<column>: <severity>: <message> [<id>]
+        match = line.strip.match(/^(?<path>.+):(?<line>\d+):(?<column>\d+): (?<severity>[^:]+): (?<message>.+) \[(?<id>[^\[]+)\]$/)
+        if match
+          issues << Issue.new(
+            path: relative_path(match[:path]),
+            location: Location.new(start_line: match[:line], start_column: match[:column]),
+            id: match[:id],
+            message: match[:message],
+            object: {
+              severity: match[:severity],
+            },
+            schema: Schema.issue,
+          )
+        end
+      end
+
+      issues
+    end
+
+    def option_include_path
+      includes = config_linter[:'include-path'] || find_paths_containing_headers
+      includes.map { |v| "-I" + v }
+    end
+
+    def find_paths_containing_headers
+      current_dir.glob(GLOB_HEADERS, File::FNM_EXTGLOB | File::FNM_CASEFOLD)
+        .select { |path| path.file? }
+        .map { |path| relative_path(path.parent, from: current_dir).to_path }
```

Comment:
[suggestion] Maybe this refactoring may be better (but I don't try it locally...):

```suggestion
      Pathname.glob(GLOB_HEADERS, File::FNM_EXTGLOB | File::FNM_CASEFOLD)
        .filter_map { |path| path.parent.to_path if path.file? }
```

See https://ruby-doc.org/core/Enumerable.html#filter_map-method

Reply:
LGTM. The `.select` and `.map` are converted to `.filter_map`.
commit e62a5d4236988b1474601bd97eb1d080d732659c

The `relative_path` method is left unchanged. Users need to set relative paths to `include-path:'. We should set the same format as users so that we can test the clang-tidy's `-I` option. Also users may get confused about absolute paths on the Sider console log.

Reply:
Maybe `Pathname.glob(GLOB_HEADERS)` should return relative paths, so it seems needless to use `relative_path`. 🤔 Right?

```console
$ irb
irb(main):001:0> Pathname.glob("*.md")
=> [#<Pathname:CHANGELOG.md>, #<Pathname:README.md>, #<Pathname:CONTRIBUTING.md>]
```

Reply:
We have an instance of `Pathname` which points `/tmp/blablabla`.

```shell
/tmp$ mkdir -p aaa/bbb/ccc/
/tmp$ touch aaa/bbb/ccc/test.md
/tmp$ cd
~$ irb
irb(main):001:0> path = Pathname("/tmp/aaa")
irb(main):002:0> path.glob("*.md")
=> []
irb(main):003:0> path.glob("**/*.md")
=> [#<Pathname:/tmp/aaa/bbb/ccc/test.md>]
irb(main):004:0>
```

Reply:
Ohhh, I missed `Pathname.glob()` in your suggestion.

Reply:
Fixed. commit c142082debb6e8e4ff1a0143c94519647018c232